Remove restriction that SpanContext must be a final class, specify creation. (#969)
open-telemetry/opentelemetry-specification: specification/trace/api.md

systems to participate in the same trace. It is fully described in the [W3C Trace Context
specification](https://www.w3.org/TR/trace-context/#tracestate-header).

> The API must implement methods to create a `SpanContext`. These methods SHOULD be the only way to

Reviewer: ```suggestion
The API MUST implement methods to create a `SpanContext`. These methods SHOULD be the only way to
```

systems to participate in the same trace. It is fully described in the [W3C Trace Context
specification](https://www.w3.org/TR/trace-context/#tracestate-header).

> The API must provide methods to create `SpanContext`. There are three ways to create
> a `SpanContext`

Reviewer: ```suggestion
The API MUST implement methods to create `SpanContext` which MUST NOT require any other API/SDK object (i.e. they SHOULD be provided as free functions, or as static functions/constructors in the ` SpanContext` interface/class). The produced SpanContext implementation when no API implementation is available MUST be interchangeable with the SpanContext produced after installing any SDK.

There are three ways to create a `SpanContext`:
```

or, what I would prefer but it's probably too late for that rather large change:

```suggestion
The API MUST provide methods to create a `SpanContext` object on the `Tracer`. This MUST be the only way to create a `SpanContext`.

There are three ways to create a `SpanContext`:
```

> - By generating a new `SpanId` and `TraceId` when there is no in-process or remote parent.

Reviewer: There should be a way to provide TraceFlags and TraceState for root spans too.

Author: Thanks - yeah I added the point about `TraceFlags`, I don't think we currently support `TraceState`. I'd like to keep this PR to reflect what we currently do if it's ok. I'd also like to avoid getting into too much detail about implementation, for example, `SamplingContext.newForSampler(Sampler)` and generate all IDs / trace flags and then call `SpanContext.of(...)` are equivalent IMO as far as the spec is concerned. I have found a general theme of over-specifying implementation detail in the spec even though it's too different based on language idioms.

Reviewer: > I don't think we currently support `TraceState`

We absolutely do. @iNikem added some descriptions how a tracestate can be "modifified" recently (the actual object instances are immutable but you can create a modified copy). Of course there must be some way of associating a tracestate with a spancontext too.

> - By generating a new `SpanId` and copying the `TraceId`, `TraceFlags`, and `TraceState` from

Reviewer: There must definitely be a way to change the TraceFlags (sampled) at this point. I think there should also be a way to change the TraceState. Maybe we could cover this by adding an additional mode of creation that accepts only a TraceState and leaves everything else (including the "remoteness") unchanged.

Author: Yup makes sense. I think https://github.com/open-telemetry/opentelemetry-specification/pull/988 addresses being able to mutate `TraceState`, if it's ok want to keep this reflecting the current state.

Reviewer: #998 would allow the sampler to choose a tracestate, but the SDK would then need to construct a new SpanContext with that tracestate, which will only work if there is an option to create a SpanContext with a tracestate. I also want to keep reflecting the current state, which is, for example in Java, that you have the option (actually the obligation) to pass a tracestate whether the spancontext is remote or local.

systems to participate in the same trace. It is fully described in the [W3C Trace Context
specification](https://www.w3.org/TR/trace-context/#tracestate-header).

The API must implement methods to create a `SpanContext`. These methods SHOULD be the only way to
create a `SpanContext`:

> - By generating a new `SpanId`, `TraceId`, and `TraceFlags` when there is no in-process or remote parent.

Reviewer: Generating the span ID must not be the responsibility of the API, see your own #896. Maybe you wanted to specify that these values should be passed as parameters? Right now it sounds like the SpanContext would generate them itself.

> - By generating a new `SpanId` and `TraceFlags` and copying the `TraceId` and `TraceState` from

Reviewer: How should SpanContext know how to generate the TraceFlags?

an in-process parent.
- By accepting values for the `SpanId`, `TraceId`, `TraceFlags`, and `TraceState` for a remote

> parent. In this case, the `SpanId`, `TraceId`, and `TraceFlags` must be used as is, and
> `TraceState` may be modified based on the rules for modification described below.

Reviewer: What is meant by "must be used as is"? In the other cases, the API will also not modify anything, right? And the span/trace ID and especially the TraceFlags may not be available in exactly the format SpanContext expects them on the incoming protocol (e.g. jaeger propagation). I suggest to remove this sentence.

```suggestion
parent.
```
